File: src/tools/get-maintenance-work-orders.ts
/**
 * Herramienta MCP para obtener órdenes de trabajo de mantenimiento de Track HS
 */

import { BaseTrackHSTool } from '../core/base-tool.js';
import { 
  MaintenanceWorkOrdersResponse, 
  GetMaintenanceWorkOrdersParams 
} from '../types/maintenance-work-orders.js';

export class GetMaintenanceWorkOrdersTool extends BaseTrackHSTool {
  name = 'get_maintenance_work_orders';
  description = 'Retrieve paginated collection of maintenance work orders from Track HS';
  
  inputSchema = {
    type: 'object' as const,
    properties: {
      updatedSince: {
        type: 'string',
        format: 'date',
        description: 'Filter by Date Time in ISO 8601 format. Use this to find recent updates to refresh your cache'
      },
      page: {
        type: 'number',
        description: 'Page Number',
        minimum: 1
      },
      size: {
        type: 'number',
        description: 'Page size, defaults to 25. No max currently but we may consider setting it at 100 or so.',
        minimum: 1,
        maximum: 100
      },
      sortColumn: {
        type: 'string',
        enum: ['id', 'scheduledAt', 'status', 'priority', 'dateReceived', 'unitId', 'vendorId', 'userId', 'summary'],
        default: 'id',
        description: 'Only id and name supported.'
      },
      sortDirection: {
        type: 'string',
        enum: ['asc', 'desc'],
        default: 'asc',
        description: 'asc or desc'
      },
      search: {
        type: 'string',
        description: 'Search with id = number or summary, unit.name, vendor.name or user.name = string.'
      },
      isScheduled: {
        type: 'number',
        enum: [1, 0],
        description: 'Filter by scheduled'
      },
      unitId: {
        type: 'string',
        description: 'Filter by unitId csv'
      },
      userId: {
        type: 'array',
        items: { type: 'number' },
        description: 'Filter by userId'
      },
      nodeId: {
        type: 'number',
        description: 'Filter by nodeId'
      },
      roleId: {
        type: 'number',
        description: 'Filter by roleId'
      },
      ownerId: {
        type: 'number',
        description: 'Filter by ownerId'
      },
      priority: {
        type: 'array',
        items: { type: 'number' },
        description: 'Filter by cleanTypeId integers'
      },
      reservationId: {
        type: 'number',
        description: 'Filter by reservationId'
      },
      vendorId: {
        type: 'number',
        description: 'Filter by vendorId'
      },
      status: {
        type: 'array',
        items: {
          type: 'string',
          enum: ['open', 'not-started', 'in-progress', 'completed', 'processed', 
                 'vendor-not-start', 'vendor-assigned', 'vendor-declined', 
                 'vendor-completed', 'user-completed', 'cancelled']
        },
        description: 'Filter by status integers'
      },
      dateScheduled: {
        type: 'string',
        format: 'date',
        description: 'Filter by date scheduled as ISO 8601 format'
      },
      startDate: {
        type: 'string',
        format: 'date',
        description: 'Filter by start date as ISO 8601 format'
      },
      endDate: {
        type: 'string',
        format: 'date',
        description: 'Filter by end date as ISO 8601 format'
      },
      problems: {
        type: 'array',
        items: { type: 'number' },
        description: 'Filter by problems ids'
      }
    },
    required: []
  };

  async execute(params: GetMaintenanceWorkOrdersParams = {}): Promise<MaintenanceWorkOrdersResponse> {
    // Validar parámetros
    this.validateParams(params);

    // Construir query parameters
    const queryParams = new URLSearchParams();
    
    // Aplicar valores por defecto
    const page = params.page || 1;
    const size = params.size || 25;
    const sortColumn = params.sortColumn || 'id';
    const sortDirection = params.sortDirection || 'asc';

    // Agregar parámetros básicos
    queryParams.append('page', page.toString());
    queryParams.append('size', size.toString());
    queryParams.append('sortColumn', sortColumn);
    queryParams.append('sortDirection', sortDirection);

    // Agregar parámetros opcionales
    if (params.updatedSince) {
      queryParams.append('updatedSince', params.updatedSince);
    }
    if (params.search) {
      queryParams.append('search', params.search);
    }
    if (params.isScheduled !== undefined) {
      queryParams.append('isScheduled', params.isScheduled.toString());
    }
    if (params.unitId) {
      queryParams.append('unitId', params.unitId);
    }
    if (params.userId && params.userId.length > 0) {
      params.userId.forEach(id => queryParams.append('userId', id.toString()));
    }
    if (params.nodeId) {
      queryParams.append('nodeId', params.nodeId.toString());
    }
    if (params.roleId) {
      queryParams.append('roleId', params.roleId.toString());
    }
    if (params.ownerId) {
      queryParams.append('ownerId', params.ownerId.toString());
    }
    if (params.priority && params.priority.length > 0) {
      params.priority.forEach(p => queryParams.append('priority', p.toString()));
    }
    if (params.reservationId) {
      queryParams.append('reservationId', params.reservationId.toString());
    }
    if (params.vendorId) {
      queryParams.append('vendorId', params.vendorId.toString());
    }
    if (params.status && params.status.length > 0) {
      params.status.forEach(s => queryParams.append('status', s));
    }
    if (params.dateScheduled) {
      queryParams.append('dateScheduled', params.dateScheduled);
    }
    if (params.startDate) {
      queryParams.append('startDate', params.startDate);
    }
    if (params.endDate) {
      queryParams.append('endDate', params.endDate);
    }
    if (params.problems && params.problems.length > 0) {
      params.problems.forEach(p => queryParams.append('problems', p.toString()));
    }

    const endpoint = `/pms/maintenance/work-orders?${queryParams.toString()}`;
    
    try {
      const result = await this.apiClient.get<MaintenanceWorkOrdersResponse>(endpoint);
      return result;
    } catch (error) {
      throw new Error(`Error al obtener órdenes de trabajo de mantenimiento: ${error instanceof Error ? error.message : 'Error desconocido'}`);
    }
  }
}

File: src/trackhs_mcp/tools/get_maintenance_work_orders.py
"""
Herramienta MCP para obtener órdenes de trabajo de mantenimiento de Track HS API
"""

from typing import Optional
from ..core.api_client import TrackHSApiClient

def register_get_maintenance_work_orders(mcp, api_client: TrackHSApiClient):
    """Registra la herramienta get_maintenance_work_orders"""
    
    @mcp.tool()
    async def get_maintenance_work_orders(
        page: int = 1,
        size: int = 10,
        sort_column: str = "id",
        sort_direction: str = "asc",
        search: str = None,
        updated_since: str = None,
        is_scheduled: int = None,
        unit_id: str = None,
        user_id: str = None,
        node_id: int = None,
        role_id: int = None,
        owner_id: int = None,
        priority: str = None,
        reservation_id: int = None,
        vendor_id: int = None,
        status: str = None,
        date_scheduled: str = None,
        start_date: str = None,
        end_date: str = None,
        problems: str = None
    ):
        """
        Get maintenance work orders from Track HS
        
        Args:
            page: Número de página (default: 1)
            size: Tamaño de página (default: 10)
            sort_column: Columna para ordenar (default: "id")
            sort_direction: Dirección de ordenamiento (default: "asc")
            search: Término de búsqueda (opcional)
            updated_since: Filtro por fecha de actualización (opcional)
            is_scheduled: Si está programado (opcional)
            unit_id: ID de la unidad (opcional)
            user_id: ID del usuario (opcional)
            node_id: ID del nodo (opcional)
            role_id: ID del rol (opcional)
            owner_id: ID del propietario (opcional)
            priority: Prioridad (opcional)
            reservation_id: ID de la reserva (opcional)
            vendor_id: ID del proveedor (opcional)
            status: Estado (opcional)
            date_scheduled: Fecha programada (opcional)
            start_date: Fecha de inicio (opcional)
            end_date: Fecha de fin (opcional)
            problems: Problemas (opcional)
        """
        query_params = {
            "page": page,
            "size": size,
            "sortColumn": sort_column,
            "sortDirection": sort_direction
        }
        
        if search:
            query_params["search"] = search
        if updated_since:
            query_params["updatedSince"] = updated_since
        if is_scheduled is not None:
            query_params["isScheduled"] = is_scheduled
        if unit_id:
            query_params["unitId"] = unit_id
        if user_id:
            query_params["userId"] = user_id
        if node_id:
            query_params["nodeId"] = node_id
        if role_id:
            query_params["roleId"] = role_id
        if owner_id:
            query_params["ownerId"] = owner_id
        if priority:
            query_params["priority"] = priority
        if reservation_id:
            query_params["reservationId"] = reservation_id
        if vendor_id:
            query_params["vendorId"] = vendor_id
        if status:
            query_params["status"] = status
        if date_scheduled:
            query_params["dateScheduled"] = date_scheduled
        if start_date:
            query_params["startDate"] = start_date
        if end_date:
            query_params["endDate"] = end_date
        if problems:
            query_params["problems"] = problems
        
        endpoint = f"/maintenance/work-orders"
        query_string = "&".join([f"{k}={v}" for k, v in query_params.items()])
        if query_string:
            endpoint += f"?{query_string}"
        
        try:
            result = await api_client.get(endpoint)
            return result
        except Exception as e:
            return {"error": f"Error al obtener órdenes de trabajo: {str(e)}"}
